pkg/process/monitor/process_monitor.go

```diff
@@ -0,0 +1,247 @@
+		if c.Regex.MatchString(pname) {
+			p.enqueueCallback(c, pid)
+		}
+	case MAPFILE:
+		mmaps, err := proc.MemoryMaps(true)
+		if err != nil {
+			log.Errorf("process %d/smaps parsing failed %s", pid, err)
+			return
+		}
+		if mmaps == nil {
+			return
+		}
+		for _, mmap := range *mmaps {
+			if c.Regex.MatchString(mmap.Path) {
+				p.enqueueCallback(c, pid)
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Subscribe(callback *ProcessCallback) (UnSubscribe func()) {
+	pm.m.Lock()
+	defer pm.m.Unlock()
+
+	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+
+	return func() {
+		pm.m.Lock()
+		defer pm.m.Unlock()
+
+		for i, c := range pm.procEventCallbacks[callback.Event] {
+			if c == callback {
+				l := len(pm.procEventCallbacks[callback.Event])
+				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
+				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
+			}
+		}
+	}
+}
+
+func (pm *ProcessMonitor) Stop() {
+	close(pm.callbackRunnerDone)
+	close(pm.done)
+	pm.wg.Wait()
+	processMonitor = nil
+}
+
+func GetProcessMonitor(cfg *config.Config) *ProcessMonitor {
+	// we want only one instance as
+	// netlink.ProcEventMonitor() netlink doesn't support multiple connections
+	if processMonitor != nil {
+		return processMonitor
+	}
+
+	p := &ProcessMonitor{}
+	p.state = STOPPED
+	p.events = make(chan netlink.ProcEvent, processMonitorMaxEvents)
+	p.done = make(chan struct{})
+	p.errors = make(chan error)
+	p.procEventCallbacks = make(map[ProcessEventType][]*ProcessCallback)
+
+	if err := netlink.ProcEventMonitor(p.events, p.done, p.errors); err != nil {
+		log.Errorf("could not create process monitor: %s", err)
+		return nil
+	}
+
+	p.callbackRunnerDone = make(chan struct{}, runtime.NumCPU())
+	p.callbackRunner = make(chan func(), runtime.NumCPU())
+	for i := 0; i < runtime.NumCPU(); i++ {
+		p.wg.Add(1)
+		go func() {
+			defer p.wg.Done()
+			for {
+				select {
+				case <-p.callbackRunnerDone:
+					return
+				case call, ok := <-p.callbackRunner:
+					if !ok {
+						continue
+					}
+					call()
+				}
+			}
+		}()
+	}
+
+	p.wg.Add(1)
+	go func() {
+		defer func() {
+			log.Info("netlink process monitor ended")
+			p.wg.Done()
+		}()
+		for {
+			select {
+			case <-p.done:
+				return
+
+			case event, ok := <-p.events:
+				if !ok {
+					return
+				}
+				if p.state == STOPPED {
+					continue
+				}
+
+				switch ev := event.Msg.(type) {
+				case *netlink.ExecProcEvent:
+					p.m.Lock()
+					for _, c := range p.procEventCallbacks[EXEC] {
+						p.evalEXECCallback(c, ev.ProcessPid)
+					}
+					p.m.Unlock()
+				case *netlink.ExitProcEvent:
+					p.m.Lock()
+					for _, c := range p.procEventCallbacks[EXIT] {
+						p.enqueueCallback(c, ev.ProcessPid)
+					}
+					p.m.Unlock()
+				}
+
+			case err, ok := <-p.errors:
+				if !ok {
+					return
+				}
+				log.Errorf("process montior error: %s", err)
+				processMonitor = nil
```

Comment:
1. why are you referring to `processMonitor` rather than `p`?
2. that's should be done under lock
3. why don't you call `Stop`? on such scenarios we are leaking the channels